test/common/http/http1/codec_impl_test.cc

```diff
@@ -1423,6 +1423,59 @@ TEST_F(Http1ServerConnectionImplTest, HeaderOnlyResponseTrainProperHeaders) {
             output);
 }
 
+TEST_F(Http1ServerConnectionImplTest, 304ResponseTransferEncodingNotAddedWhenContentLengthPresent) {
+  initialize();
+
+  NiceMock<MockRequestDecoder> decoder;
+  Http::ResponseEncoder* response_encoder = nullptr;
+  EXPECT_CALL(callbacks_, newStream(_, _))
+      .WillOnce(Invoke([&](ResponseEncoder& encoder, bool) -> RequestDecoder& {
+        response_encoder = &encoder;
+        return decoder;
+      }));
+
+  Buffer::OwnedImpl buffer("GET / HTTP/1.1\r\nif-none-match: \"1234567890\"\r\n\r\n");
+  auto status = codec_->dispatch(buffer);
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(0U, buffer.length());
+
+  std::string output;
+  ON_CALL(connection_, write(_, _)).WillByDefault(AddBufferToString(&output));
+
+  TestResponseHeaderMapImpl headers{
+      {":status", "304"}, {"etag", "\"1234567890\""}, {"content-length", "123"}};
+  response_encoder->encodeHeaders(headers, true);
+  EXPECT_EQ("HTTP/1.1 304 Not Modified\r\netag: \"1234567890\"\r\ncontent-length: 123\r\n\r\n",
```

Comment:
Preservation of content-length in 304 responses worries me a bit since it may allow for desync between server, intermediate proxies and client in cases where one of these doesn't interpret content-length in 304 responses correctly.  I guess this behavior is not changing in this PR?

Reply:
Correct. I didn't change the behavior but only make a test to confirm/validate that behavior. The newer Http spec allows Content-Length to be in the 304 response. While the older spec didn't explicitly prohibit it, it says 304 response should only have the list of headers which never included Content-Length. I have seen some newer software starts adding Content-Length to 304 response. 

This concern actually apply to Transfer-Encoding as well, some client may try to read the chunked body which is never there for a 304 response and that's the main reason I am changing this.

Reply:
Can we actually sets mocks on the decoder to make sure http_parser handles the content-length-without-body correctly, sending headers and end_stream up?

Reply:
I have not setup mock but I did add tests to do 2 requests back to back with the first upstream response sending content-length header with no body and made sure the 2nd request did not get error from upstream and verify it's getting the correct response.

Reply:
The test is in protocol_integration_test.cc